### PR TITLE
Rebased develop with NUMA bindings and proposed upstream PRs

### PR DESCRIPTION
**Pull request by byrnedj**
This contains the following proposed upstream patches:
-   Introduce 'markedForEviction' state for the Item, https://github.com/facebook/CacheLib/pull/183
-   Add combined locking support for MMContainer, https://github.com/facebook/CacheLib/pull/182
-   Expose 'eviction' and 'moving' states for an item, https://github.com/facebook/CacheLib/pull/172/commits/6aa40781fafba72dc55bd25b324c63f39a89f406
-   Change the cache size calculation to use getCacheSize(), https://github.com/intel/CacheLib/pull/48
-   Add memory tier configs to cache allocator based on NUMA bindings, https://github.com/intel/CacheLib/pull/49
-   CompressedPtr, https://github.com/facebook/CacheLib/pull/188
-   Add memory usage statistics for allocation classes, https://github.com/intel/CacheLib/pull/42
     - RollingStats per class, https://github.com/intel/CacheLib/pull/47

The rest of develop features have been rebased:
 - AC stats - approx slab percentage per class
 - background  workers
 - transparent item sync - with waiters
 - configs for multi-tier now rely on NUMA bindings
 - all CI related commits have been squashed together (first commit 5130a658223973cb97d610a3a3d08b8249aa2352)
 
I did not include file backed shared memory in this version. 

All tests compile and pass and there is no performance difference in the hit_ratio configs. Navy consistency runs and benchmarks work.

I did my best to separate out the multi-tier additional patches for the upstream PRs (i.e. rolling stats upstream commit is followed by the multi-tier implementation for that feature - such as adding TierId and outputting)

- - -
This change is [Reviewable](https://reviewable.io/reviews/intel/CacheLib/56)

### CODE REVIEW

**Comment by guptask** [comment]
What's the overall goal with this rebase:
Is it to bring the latest changes from Meta into develop ?
OR
Did you refactor the feature commits so that we can just cherry-pick one or more commit from here and raise upstream PRs ? If this is the primary goal, then I think it would help if you group the dependent features together (grouped based on the staggered upstreaming plan) e.g. rolling stats cannot exist independently. It needs AC stats to work.

**Comment by byrnedj** [comment]
> Is it to bring the latest changes from Meta into develop ?
Yes 
> Did you refactor the feature commits so that we can just cherry-pick one or more commit from here and raise upstream PRs ? If this is the primary goal, then I think it would help if you modify the PR description to group the dependent features together (grouped based on the staggered upstreaming plan) e.g. rolling stats cannot exist independently. It needs AC stats to work.
This is also the goal of the rebase - I put rolling stats as a sub bullet.

**Comment by igchor** [comment]
I'm not sure if we should have this commit (Enable touchValue by default): https://github.com/intel/CacheLib/pull/56/commits/08916e3c11a5c9a156f290f529ac27aca332293a I think it's fine to remove it and just set it in the config every time.

**Comment by igchor** [comment]
Why does this commit (added per tier pool class rolling average latency (based on upstream PR)) https://github.com/intel/CacheLib/pull/56/commits/6b41f3bb0f330d2dcf9549194d7f430e523c53f8 introduce kMaxTiers? Isn't it added in some earlier commit?

**Comment by guptask** [comment]
> Why does this commit (added per tier pool class rolling average latency (based on upstream PR)) [6b41f3b](https://github.com/intel/CacheLib/commit/6b41f3bb0f330d2dcf9549194d7f430e523c53f8) introduce kMaxTiers? Isn't it added in some earlier commit?

kMaxTiers change should be part of compressed ptr commit since by design those changes support <= 2 tiers. It is currently not part of https://github.com/facebook/CacheLib/pull/188

**Comment by igchor** [comment]
> > Why does this commit (added per tier pool class rolling average latency (based on upstream PR)) [6b41f3b](https://github.com/intel/CacheLib/commit/6b41f3bb0f330d2dcf9549194d7f430e523c53f8) introduce kMaxTiers? Isn't it added in some earlier commit?
> 
> kMaxTiers change should be part of compressed ptr commit since by design those changes support <= 2 tiers. It is currently not part of [facebook#188](https://github.com/facebook/CacheLib/pull/188)

Ok, I just thought we introduced this constant in some earlier commit (on develop).

File: cachelib/allocator/Refcount.h [hunk 1]
```diff
@@ -134,30 +130,47 @@ class FOLLY_PACK_ATTR RefcountWithFlags {
...
+     };
+
+     auto newValue = [](const Value curValue) {
+       return (curValue + static_cast<Value>(1));
+     };
+
+     if (atomicUpdateValue(predicate, newValue)) {
+         return incOk;
+     } 
+     Value val = __atomic_load_n(&refCount_, __ATOMIC_RELAXED);
```
Comment (igchor) [inline cachelib/allocator/Refcount.h:165]: Hm, this part of the code looks incorrect. There is a race possible. We should probably return the incResult from predicate (through reference perhaps).

Is this a rebase conflict?

File: cachelib/allocator/tests/MMTypeTest.h [hunk 2]
```diff
@@ -231,26 +234,31 @@ void MMTypeTest<MMType>::testRemoveBasic(Config config) {
...
-  for (auto iter = c.getEvictionIterator(); iter;) {
-    auto& node = *iter;
-    ASSERT_TRUE(node.isInMMContainer());
-
-    // this will move the iter.
-    c.remove(iter);
-    ASSERT_FALSE(node.isInMMContainer());
-    if (iter) {
-      ASSERT_NE((*iter).getId(), node.getId());
+  c.withEvictionIterator([&foundNodes, &c](auto&& iter) {
```
Comment (igchor) [inline cachelib/allocator/tests/MMTypeTest.h:249]: I think this is based on old version of CS-patch PR. In the newest one I have rolled-back those changes to tests to limit number of lines changed.  You can see here (https://github.com/facebook/CacheLib/pull/182/files#diff-5298053236abd39332070ce33eda4db9571ef064728bfe48bbf1747306c32997) that I just extended the existing tests with verifyIterationVariants

**Comment by igchor** [comment]
Where is the fix in (Fix moveRegularItemWithSync and add tests): https://github.com/intel/CacheLib/pull/56/commits/2d80fb1c0d5216e44e894228120f6c2c5981e568 ? I only see tests in that commit

**Comment by igchor** [comment]
I'm not sure if this commit (Support for approx free percentage in each allocation class based): https://github.com/intel/CacheLib/pull/56/commits/e4ce2e73858d83e6eefcdd171d72c817a697723b is still needed after: https://github.com/intel/CacheLib/pull/56/commits/e756b242f7dc51b93fb54dcf9596471ba3d371e7 

I think it would be better to just leave the commit from https://github.com/intel/CacheLib/pull/42/commits and use that in background workers (functionality is basically the same). What do you think?

**Comment by igchor** [comment]
https://github.com/intel/CacheLib/pull/56/commits/7d105a75741f13f4f5ac197b25f04fa0764096a8 (Create token before marking item as exclusive) can be part of the CS-patch commit

**Comment by igchor** [comment]
I'm not sure if we need to keep https://github.com/intel/CacheLib/pull/56/commits/2d80fb1c0d5216e44e894228120f6c2c5981e568 (Implemented async Item movement between tier) as a separate commit. The final solution is to use the moving bit so perhaps we should squash with critical section patch?

**Comment by guptask** [comment]
*[`cachelib/allocator/memory/CompressedPtr.h` line 151 at r1](https://reviewable.io/reviews/intel/CacheLib/56#-NOVz_tL2qe0_Ym8Vola:-NOVz_tMBnnDlJfpANwZ:bsrdvvb) ([raw file](https://github.com/intel/CacheLib/blob/ef893e78d7b19a2babcf31362e986b1cd314bb73/cachelib/allocator/memory/CompressedPtr.h#L151)):*
> ```objc
>   uint32_t getAllocIdx(bool isMultiTiered) const noexcept {
>     XDCHECK(!isNull());
>     auto noTierIdPtr = isMultiTiered ? ptr_ & ~kTierIdxMask : ptr_;
> ```

This section has changed based on latest review comments from Meta. Please refer to https://github.com/facebook/CacheLib/pull/188/commits/969d5c91c00fafc50ef200e005edafed7e4bc92d

**Comment by guptask** [comment]
*[`cachelib/allocator/memory/CompressedPtr.h` line 46 at r1](https://reviewable.io/reviews/intel/CacheLib/56#-NOVzj1n5WuE2I6W5pBJ:-NOVzj1n5WuE2I6W5pBK:b-rvewpa) ([raw file](https://github.com/intel/CacheLib/blob/ef893e78d7b19a2babcf31362e986b1cd314bb73/cachelib/allocator/memory/CompressedPtr.h#L46)):*
> ```objc
> // the  slab  index. Hence we can index 128 GiB of memory in slabs per tier and
> // index anything more than 64 byte allocations inside the slab using a 32 bit
> // representation.
> ```

The summary has changed based on review comments from Meta. Please refer to https://github.com/facebook/CacheLib/pull/188/commits/969d5c91c00fafc50ef200e005edafed7e4bc92d

**Comment by byrnedj** [comment]
@guptask I rebased the CompressedPtr with the latest changes on the upstream PR

@igchor I used usageFraction instead of slabsApproxFree - it works but we might want to think about making pool stats multi-tier so we have per tier stats.

I have also included the fix for the race condition in the insertOrReplace during move - see 
https://github.com/intel/CacheLib/pull/56/commits/02b85e1c1cf16b79f38779e91df618045f3c44a0

File: cachelib/allocator/CacheAllocator-inl.h [hunk 3]
```diff
@@ -1526,17 +1548,33 @@ CacheAllocator<CacheTrait>::createPutToken(Item& item) {
...
+      //from NVM
+      XDCHECK(!it.isAccessible());
+      HashedKey hk{it.getKey()};
+      auto lock = nvmCache_ ? nvmCache_->getItemDestructorLock(hk)
+                            : std::unique_lock<std::mutex>();
+      if (it.isNvmClean() && !it.isNvmEvicted()) {
+        // item is to be replaced and the destructor will be executed
+        // upon memory released, mark it in nvm to avoid destructor
+        // executed from nvm
+        nvmCache_->markNvmItemRemovedLocked(hk);
```
Comment (igchor) [inline cachelib/allocator/CacheAllocator-inl.h:1567]: Do we need to call this here? Isn't the call in `insertOrReplace` enough?

Comment (byrnedj) [reply cachelib/allocator/CacheAllocator-inl.h]: Yes as we discussed - I fixed in latest commit - https://github.com/intel/CacheLib/pull/56/commits/5eb049e97fa24e8bced785e392ebc2437d0f4b25

**Pull request closed without merging**